### ISSUE TITLE
Add a 50cm-1 lineshape truncation by default

### ISSUE THREAD

**Issue by erwanp** [radis/radis#340]
### 🎯 Describe what you'd like

Since 0.9.20 (Aug. 2019  https://github.com/radis/radis/pull/39  RADIS has been using the DIT  (or DLM , we're not set on the name yet!) method as the default method to compute spectra, with a FFT algorithm to resolve the convolutions. The DIT+FFT approach has no lineshape truncatino, which is a good thing for small spectra, where legacy LBL codes usually rely on a truncation of 5, 10 or 50 cm-1 to improve computational performance at the cost of accuracy.

However, in the far wings (few hundreds of cm-1) lineshapes exhibit a sub-lorentzian behavior : therefore, for large spectra, having no truncation is less physically accurate than having a truncation (see discussion on [💬 Slack][https://radis-radiation.slack.com/archives/C01JBV54G11/p1610183178001400) )

Here, we recommend to use a default truncation of 300 cm-1. More accurate, molecule-dependant truncations may be implemented later.

### 💡 Possible implementations

It is hard to implement a truncation with a FFT algorithm (discussion with @dcmvdbekerom for details).  However, a truncatino in wavenumber space is easy. RADIS already has a DIT+analytical Voigt (in wavenumber space) implement ; it's actually the first implementation that was done #37. Until this version, this version was lagging behind DIT+FFT in terms of performance ; making it unpractical for large number of lines. Recent change #323  made DIT+Voigt competitive; and even faster than DIT+FFT in most cases (see the extensive benchmarks done by @anandxkumar in https://anandxkumar.github.io/Benchmark_Visualization_GSoC_2021 . Version post #323 is the one with `oa.convolve` ) 

The only drawback is that the analytical Voigt introduces some errors compared to real-Voigt. For small spectra, close to the line centers, we should keep the DIT+FFT approach (real Voigts). 

Suggested implementation, in `calc_spectrum` : 
- calculations with less than 300cm-1 range : keep using DIT+FFT as default
- calculations  with spectral range > 300 cm-1 : use DIT+Voigt with a 300cm-1 cutoff

### Example

A comparison of both methods using the CO2 HITEMP database (fetched automatically with #281 )   (example adapted from @anandxkumar 's benchmarks)

```python

from radis import SpectrumFactory
from radis import plot_diff

# %% Condition Parameters
wmin = 500
wmax = 4000
dv = 0.01
T = 300.0 #K
p = 1 #bar
broadening_max_width=300    # lineshape broadening width, ext​

sf = SpectrumFactory(wavenum_min=wmin, wavenum_max=wmax, 
                  pressure=p,
                  isotope='1,2,3',
                  molecule='CO2',
                  wstep=dv,
                  broadening_max_width=broadening_max_width, 
                  cutoff=0, # 1e-27,
                  verbose=3,
                  )

sf.fetch_databank('hitemp')

#%% Calculate DIT + FFT
sf.params['optimization'] = 'simple'
sf.params['broadening_method'] = 'fft'

s_ditfft = sf.eq_spectrum(T, name = "DIT+FFT")

#%% Voigt 
sf.params['optimization'] = 'simple'
sf.params['broadening_method'] = 'voigt'

s_ditvoigt = sf.eq_spectrum(T, name = f"DIT+Voigt ({broadening_max_width:.0f}"+r"cm$^{-1}$ trunc)")


# %% Compare

s_ditfft.print_perf_profile()
s_ditvoigt.print_perf_profile()
fig, [ax0, ax1] = plot_diff(s_ditfft, s_ditvoigt, "abscoeff")
ax0.set_yscale("log")
ax0.set_ylim(ymin=1e-13, ymax=1e3)
ax1.set_ylim(ymin=-0.0025, ymax=0.0025)

assert s_ditfft.conditions["lines_calculated"] == s_ditvoigt.conditions["lines_calculated"]
print(s_ditfft.conditions["lines_calculated"], "lines calculated")
```

![image](https://user-images.githubusercontent.com/16088743/128903860-136c3827-2126-4b6b-8cf4-685837f3029a.png)

Black is current default ; red would be the new default 
New default is 4x faster on my laptop (1 CPU)

DIT+FFT profiler :
 <details>
<summary>   spectrum_calculation      46.249s ████████████████</summary>
        check_line_databank             0.137s 
        reinitialize                    1.055s 
            copy_database                   1.044s 
            memory_usage_warning            0.010s 
            reset_population                0.000s 
            others                          0.001s 
        scaled_eq_linestrength          0.835s 
        calc_lineshift                  0.075s 
        calc_hwhm                       0.671s 
        generate_wavenumber_arrays      0.055s 
        calc_line_broadening            43.329s ██████████████
            precompute_DLM_lineshapes       2.932s █
            DLM_Initialized_vectors         0.000s 
            DLM_closest_matching_line       1.484s 
            DLM_Distribute_lines            0.858s 
            DLM_convolve                    37.883s █████████████
            others                          0.172s 
        calc_other_spectral_quan        0.070s 
        generate_spectrum_obj           0.012s 
        others                          0.185s 
</details>

DIT+Voigt (300cm$^\mathrm{-1}$ trunc) profiler :
<details>
<summary>
    spectrum_calculation      12.433s ████████████████</summary>
        check_line_databank             0.104s 
        reinitialize                    1.147s █
            copy_database                   1.144s █
            memory_usage_warning            0.003s 
            reset_population                0.000s 
        scaled_eq_linestrength          0.772s 
        calc_lineshift                  0.077s 
        calc_hwhm                       1.265s █
        generate_wavenumber_arrays      0.038s 
        calc_line_broadening            8.930s ███████████
            precompute_DLM_lineshapes       0.606s 
            DLM_Initialized_vectors         0.000s 
            DLM_closest_matching_line       1.453s █
            DLM_Distribute_lines            0.879s █
            DLM_convolve                    5.876s ███████
            others                          0.116s 
        calc_other_spectral_quan        0.071s 
        generate_spectrum_obj           0.012s 
        others                          0.135s 
</details>

7,107,089 lines calculated             (>>> 0.5  and   2e11 points*lines/s 🥇  ) 

---
<sub>*Tip : you can vote for [New Features](https://github.com/radis/radis/issues/79) on https://feathub.com/radis/radis*</sub>

### References
- Physics for default 50 cm-1: Discussed in the `features-advanced_lineshapes` channel (https://radis-radiation.slack.com/archives/C01JBV54G11/p1629636709000900)
- some on-going work on characterizing physical truncations : https://www.lmd.jussieu.fr/~mturbet/research_spectroscopy.html

**Comment by dcmvdbekerom** [comment]
>Suggested implementation, in calc_spectrum :
>calculations with less than 300cm-1 range : keep using DIT+FFT as default
>calculations with spectral range > 300 cm-1 : use DIT+Voigt with a 300cm-1 cutoff

I would even go so far as to use DIT+Voigt with truncation as a default _always_.

Until we introduced our algorithm, approximated Voigt's is all people ever used, so I imagine that in all but a few select cases Voigt + DIT will be fine. I think it is preferable that the code is predictable, meaning it doesn't suddenly change method below a seemingly arbitrary spectral range, than trying to cater to everyone. 

Of course we can keep the DIT+FFT in RADIS as an option in case people really need it, but I doubt people will ever be so desperate for an extra 1% of accuracy or so.

This might feel a bit uncomfortable since the exact profiles were one of our selling points, but the way I see it the road to progress is never a straight one and if we have to take one step back to make two steps forward so be it.

**Comment by erwanp** [comment]
I base this from Fig.9 of the DIT Paper. Updated below : I expect performance of DIT/Voigt to be that of DIT/FFT; but accuracy will remain close to Voigt 

![image](https://user-images.githubusercontent.com/16088743/129291826-f2a07b29-9331-43cd-8164-da50f564a896.png)


Ok for using a single methods  (mentioning @anandxkumar  as it will impact #343 )  

The only thing we need; then; is to split the `broadening_max_width` parameter into a Truncation (will be 300 cm-1) and a include_neighbour_lines parameter (will be few cm-1; maybe current default of 15 cm-1; maybe even 0 by default + an a-posteriori warning based on the average width of the computed lines) . 
For very low resolution spectra of 1-2 cm-1 it's a huge penalty )

**Comment by dcmvdbekerom** [comment]
>The only thing we need; then; is to split the broadening_max_width parameter into a Truncation (will be 300 cm-1) and a include_neighbour_lines parameter 

Shouldn't the include_neighbour_lines parameter really be the same width as broadening_max_width? Because that 300cm-1 is now the "reach" of a single line.

**Comment by erwanp** [comment]
Strictly speaking; yes; but at the cost of high performance, for a minimal gain in accuracy

. When fitting a laser-diode experiment you'll see a few strong lines and want to run RADIS with sub-microsecond iterations; so you can't afford the 100k+ lines in the 300 cm-1 range. Plus, your main source of error is probably the non-Voigt lineshape, not so much the contribution of neighbour lines. 

Neighbour lines is more relevant for heat-transfer in the line wings region.

**Comment by dcmvdbekerom** [comment]
The new folding correction for Lorentzians in real space (see: https://github.com/dcmvdbekerom/discrete-integral-transform/blob/master/demo/discrete_integral_transform_log.py) would allow us to truncate the lineshape when using FFT.

In light of this, I would suggest to implement an overlap-add scheme for the FFT method, which if it works we could make default. I can start working on this in a few weeks.

Also one question about the physics, why is the cutoff 300cm-1? I would expect the cutoff to be a function of the Lorentzian width. With a fixed number such as 300cm-1, it would cut off at different parts of the lineshape depending on the width, which I would think is undesirable.

**Comment by erwanp** [comment]
Great news for FFT with truncation !

The current truncation is implemented in real-space in #343 ; we'll use that for 0.9.30, and switch to the overlap-add FFT once it's ready for 1.0 ! 

About the threshold : I've been reading Hartmann's book about far-wing models. More on this on Slack #advanced_lineshapes
